File: cpp/TitanMicroBus.cpp
#include "TitanMicroBus.hpp"
#include <cstring>
#include <cstdlib>

namespace Titan {

std::map<uint16_t, std::string> TitanMicroBus::s_registers;
std::map<uint8_t, std::vector<NodeHandler>> TitanMicroBus::s_commandHandlers;
std::map<uint16_t, std::vector<RegisterListener>> TitanMicroBus::s_registerListeners;
uint8_t TitanMicroBus::s_rxBuffer[512];
size_t TitanMicroBus::s_rxHead = 0;
uint32_t TitanMicroBus::s_seqNo = 1;

void TitanMicroBus::init() {
    s_registers.clear();
    s_commandHandlers.clear();
    s_registerListeners.clear();
    s_rxHead = 0;
    s_seqNo = 1;

    // Set default core registers
    writeInt(REG_BUS_STATUS, 1); // Online
    writeInt(REG_ACTIVE_SCREEN, 10); // Home Screen
    write(REG_KEYPAD_DIAL_BUFFER, "");
}

bool TitanMicroBus::emit(uint32_t targetId, uint8_t cmd, const uint8_t* payload, uint32_t len, uint8_t flags) {
    TitanHeader hdr;
    hdr.signature = 0x5442;
    hdr.version = 0x02;
    hdr.cmdType = cmd;
    hdr.senderExt = 101; // Default Local Node ID
    hdr.targetExt = targetId;
    hdr.payloadLen = len;
    hdr.seqNo = s_seqNo++;
    hdr.sessionId = 0;
    hdr.flags = flags;
    hdr.checksum = 0xAA;

    // Internal loopback dispatch
    auto it = s_commandHandlers.find(cmd);
    if (it != s_commandHandlers.end()) {
        for (auto& handler : it->second) {
            handler(cmd, hdr.senderExt, payload, len);
        }
    }
    return true;
}

void TitanMicroBus::subscribe(uint8_t cmd, NodeHandler handler) {
    s_commandHandlers[cmd].push_back(handler);
}

void TitanMicroBus::write(uint16_t reg, const std::string& value) {
    s_registers[reg] = value;
    auto it = s_registerListeners.find(reg);
    if (it != s_registerListeners.end()) {
        for (auto& listener : it->second) {
            listener(reg, value);
        }
    }
}

void TitanMicroBus::writeInt(uint16_t reg, int32_t value) {
    write(reg, std::to_string(value));
}

std::string TitanMicroBus::read(uint16_t reg) {
    auto it = s_registers.find(reg);
    return (it != s_registers.end()) ? it->second : "";
}

int32_t TitanMicroBus::readInt(uint16_t reg, int32_t defaultValue) {
    auto it = s_registers.find(reg);
    if (it == s_registers.end() || it->second.empty()) return defaultValue;
    return std::atoi(it->second.c_str());
}

void TitanMicroBus::onRegisterChange(uint16_t reg, RegisterListener listener) {
    s_registerListeners[reg].push_back(listener);
}

void TitanMicroBus::processIncomingBytes(const uint8_t* data, size_t len) {
    if (data == nullptr || len == 0) return;
    if (s_rxHead + len > 512) s_rxHead = 0; // Prevent overflow

    std::memcpy(&s_rxBuffer[s_rxHead], data, len);
    s_rxHead += len;

    while (s_rxHead >= TITAN_HEADER_SIZE) {
        // Check magic bytes 'TB' (0x54, 0x42)
        if (s_rxBuffer[0] != 0x54 || s_rxBuffer[1] != 0x42) {
            // Shift 1 byte and realign
            std::memmove(&s_rxBuffer[0], &s_rxBuffer[1], s_rxHead - 1);
            s_rxHead--;
            continue;
        }

        uint8_t cmd = s_rxBuffer[3];
        int32_t payloadLen = (s_rxBuffer[12] << 24) | (s_rxBuffer[13] << 16) | (s_rxBuffer[14] << 8) | s_rxBuffer[15];

        if (payloadLen < 0 || payloadLen > 4096) {
            s_rxHead = 0; // Malformed packet recovery
            break;
        }

        size_t totalPacketSize = TITAN_HEADER_SIZE + payloadLen;
        if (s_rxHead < totalPacketSize) {
            break; // Awaiting remaining payload bytes
        }

        const uint8_t* payload = &s_rxBuffer[TITAN_HEADER_SIZE];
        int32_t sender = (s_rxBuffer[4] << 24) | (s_rxBuffer[5] << 16) | (s_rxBuffer[6] << 8) | s_rxBuffer[7];

        // Dispatch command
        auto it = s_commandHandlers.find(cmd);
        if (it != s_commandHandlers.end()) {
            for (auto& handler : it->second) {
                handler(cmd, sender, payload, payloadLen);
            }
        }

        // Consume frame
        size_t remaining = s_rxHead - totalPacketSize;
        if (remaining > 0) {
            std::memmove(&s_rxBuffer[0], &s_rxBuffer[totalPacketSize], remaining);
        }
        s_rxHead = remaining;
    }
}

} // namespace Titan

File: ts/TitanMicroBus.ts
'use strict';

/**
 * 🚀 TITAN MICRO-BUS (TypeScript / JavaScript Core)
 * ══════════════════════════════════════════════════════════════════
 * Ultra-fast 2-Byte (UInt16: 0 - 65535) Memory Register Highway &
 * 24-Byte Binary Packet Router for UI, DB, Audio, and Hardware.
 */

export const TITAN_SIGNATURE = 0x5442; // 'TB'
export const TITAN_VERSION = 0x02;
export const TITAN_HEADER_SIZE = 24;

export enum CMD {
  REGISTER       = 0x08,
  REGISTER_ACK   = 0x09,
  INVITE         = 0x10,
  ACCEPT         = 0x11,
  REJECT         = 0x12,
  HANGUP         = 0x13,
  AUDIO_FRAME    = 0x14,
  VIDEO_FRAME    = 0x15,
  CHAT_MESSAGE   = 0x20,
  KEYPAD_EVENT   = 0x21,
  DISPLAY_WRITE  = 0x22,
  RELAY_SET      = 0x23,
  SENSOR_STREAM  = 0x24,
  HEARTBEAT      = 0x30,
  CUSTOM_ACTION  = 0x40,
}

export type BusListener = (value: any, registerId: number) => void;
export type CmdHandler = (cmd: number, senderId: number, payload: Buffer | any) => void;

class TitanMicroBusEngine {
  private registers: Map<number, any> = new Map();
  private listeners: Map<number, Set<BusListener>> = new Map();
  private cmdHandlers: Map<number, Set<CmdHandler>> = new Map();
  private seqNo: number = 1;
  public myNodeId: number = 101;

  constructor() {
    this.initDefaults();
  }

  private initDefaults(): void {
    this.write(1, 'ONLINE'); // REG_BUS_STATUS
    this.write(10, 'Home');  // REG_ACTIVE_SCREEN
    this.write(1000, '');    // REG_KEYPAD_DIAL_BUFFER
  }

  // 1. 2-Byte Register Read / Write (0 - 65,535)
  public write(reg: number, value: any, silent = false): void {
    this.registers.set(reg, value);
    if (!silent) {
      const subs = this.listeners.get(reg);
      if (subs) {
        subs.forEach(listener => {
          try { listener(value, reg); } catch (e) { console.error(e); }
        });
      }
    }
  }

  public read(reg: number, defaultValue: any = null): any {
    return this.registers.has(reg) ? this.registers.get(reg) : defaultValue;
  }

  // 2. Register Change Subscription
  public subscribe(reg: number, listener: BusListener): () => void {
    if (!this.listeners.has(reg)) {
      this.listeners.set(reg, new Set());
    }
    this.listeners.get(reg)!.add(listener);

    // Initial trigger
    if (this.registers.has(reg)) {
      listener(this.registers.get(reg), reg);
    }

    return () => {
      this.listeners.get(reg)?.delete(listener);
    };
  }

  // 3. Command Packing & Dispatch
  public emit(targetId: number, cmd: number, payload: any = null, isJson = false): Buffer {
    let payloadBuf: Buffer;
    let flags = 0;

    if (Buffer.isBuffer(payload)) {
      payloadBuf = payload;
    } else if (typeof payload === 'string') {
      payloadBuf = Buffer.from(payload, 'utf8');
    } else if (payload !== null && typeof payload === 'object') {
      payloadBuf = Buffer.from(JSON.stringify(payload), 'utf8');
      flags = 1; // JSON mode
    } else if (typeof payload === 'number') {
      payloadBuf = Buffer.alloc(4);
      payloadBuf.writeInt32BE(payload, 0);
    } else {
      payloadBuf = Buffer.alloc(0);
    }

    const frame = Buffer.alloc(TITAN_HEADER_SIZE + payloadBuf.length);
    frame.writeUInt16BE(TITAN_SIGNATURE, 0);
    frame.writeUInt8(TITAN_VERSION, 2);
    frame.writeUInt8(cmd, 3);
    frame.writeInt32BE(this.myNodeId, 4);
    frame.writeInt32BE(targetId, 8);
    frame.writeInt32BE(payloadBuf.length, 12);
    frame.writeInt32BE(this.seqNo++, 16);
    frame.writeUInt16BE(0, 20); // Session ID
    frame.writeUInt8(flags, 22);
    frame.writeUInt8(0xAA, 23); // Checksum

    if (payloadBuf.length > 0) {
      payloadBuf.copy(frame, TITAN_HEADER_SIZE);
    }

    // Local dispatch
    const handlers = this.cmdHandlers.get(cmd);
    if (handlers) {
      handlers.forEach(h => {
        try { h(cmd, this.myNodeId, payload); } catch (e) { console.error(e); }
      });
    }

    return frame;
  }

  public onCommand(cmd: number, handler: CmdHandler): () => void {
    if (!this.cmdHandlers.has(cmd)) {
      this.cmdHandlers.set(cmd, new Set());
    }
    this.cmdHandlers.get(cmd)!.add(handler);
    return () => {
      this.cmdHandlers.get(cmd)?.delete(handler);
    };
  }

  // 4. Ingest Raw TCP Stream Bytes
  public ingestBytes(chunk: Buffer): void {
    if (chunk.length < TITAN_HEADER_SIZE) return;
    const sig = chunk.readUInt16BE(0);
    if (sig !== TITAN_SIGNATURE) return;

    const cmd = chunk.readUInt8(3);
    const sender = chunk.readInt32BE(4);
    const pLen = chunk.readInt32BE(12);
    const flags = chunk.readUInt8(22);

    let payload: any = null;
    if (pLen > 0 && chunk.length >= TITAN_HEADER_SIZE + pLen) {
      const rawPayload = chunk.subarray(TITAN_HEADER_SIZE, TITAN_HEADER_SIZE + pLen);
      if (flags === 1) {
        try { payload = JSON.parse(rawPayload.toString('utf8')); } catch { payload = rawPayload; }
      } else {
        payload = rawPayload;
      }
    }

    const handlers = this.cmdHandlers.get(cmd);
    if (handlers) {
      handlers.forEach(h => h(cmd, sender, payload));
    }
  }
}

export const TitanMicroBus = new TitanMicroBusEngine();
